WKWebViews should correctly indicate whether passkeys are available

aznair opened the pull request:
#### 0cf1f113fe36d059f9b6f5f1daa5a08c3c332805
<pre>
WKWebViews should correctly indicate whether passkeys are available
<a href="https://bugs.webkit.org/show_bug.cgi?id=250912">https://bugs.webkit.org/show_bug.cgi?id=250912</a>
rdar://104194528

Reviewed by J Pascoe.

isConditionalMediationAvailable and isUserVerifyingPlatformAuthenticatorAvailable
now check if either the client is an entitled web browser or the relying party
is in the client&apos;s associated domains and return false if neither are true.

Existing checks are performed after this new check.

To accomplish this, Document from BasicCredential is now plumbed through to
AuthenticatorCoordinator and SecurityOriginData is then passed to
WebAuthenticatorCoordinatorProxy.

* Source/WebCore/Modules/credentialmanagement/BasicCredential.cpp:
(WebCore::BasicCredential::isConditionalMediationAvailable):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::isConditionalMediationAvailable const):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.h:
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinatorClient.h:
* Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.messages.in:
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.cpp:
(WebKit::WebAuthenticatorCoordinator::isConditionalMediationAvailable):
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.h:

Canonical link: <a href="https://commits.webkit.org/259253@main">https://commits.webkit.org/259253@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6cfd12f0f161f76cee99d54daa1b2d1750dd7b2a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104425 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13504 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37333 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113641 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173933 "Failed to checkout and rebase branch from PR 8891") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108348 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14606 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4420 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96650 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112678 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110193 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11255 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94315 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38874 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93121 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25926 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/80543 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6863 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27283 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6992 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3839 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13020 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46841 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6378 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8786 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -523,14 +523,21 @@ static inline void continueAfterRequest(RetainPtr<id <ASCCredentialProtocol>> cr
 void WebAuthenticatorCoordinatorProxy::isUserVerifyingPlatformAuthenticatorAvailable(const SecurityOriginData& data, QueryCompletionHandler&& handler)
 {
-    if ([getASCWebKitSPISupportClass() shouldUseAlternateCredentialStore]) {
+    if (isConditionalMediationAvailableImpl(data)) {
```
If this returns false, is there any chance the user will be able to use the platform authenticator. It's possible we should either remove or modify the local service call below.

aznair replied:
I've done some restructuring to the methods in this file:

Helper (now named canCurrentProcessAccessPasskeysForRelyingParty per suggestion) now only checks if the SPI support class responds to the new SPI message and calls it.

isConditionalMediationAvailable
1. Call helper to see if passkey access is allowed, call handler with false if not
2. Call handler with the result of shouldUseAlternateCredentialStore

isUserVerifyingPlatformAuthenticatorAvailable
1. Call helper to see if passkey access is allowed, call handler with false if not
2. If alternate credential store is available, do the existing check of disallowed relying parties
3. Otherwise return LocalService::isAvailable()

I think this is the right way to add the new check to existing flows.

pascoej commented on Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
```diff
@@ -523,14 +523,21 @@ static inline void continueAfterRequest(RetainPtr<id <ASCCredentialProtocol>> cr
 }
 
-void WebAuthenticatorCoordinatorProxy::isConditionalMediationAvailable(QueryCompletionHandler&& handler)
+static inline bool isConditionalMediationAvailableImpl(const WebCore::SecurityOriginData& data)
```
nit: This name strikes me as odd. Maybe we can name it similar to the SPI `canCurrentProcessAccessPasskeysForRelyingParty`?

aznair replied:
Done

pascoej reviewed (approved):
LGTM

jasl23 committed:
WKWebViews should correctly indicate whether passkeys are available
https://bugs.webkit.org/show_bug.cgi?id=250912
rdar://104194528

Reviewed by J Pascoe.

isConditionalMediationAvailable and isUserVerifyingPlatformAuthenticatorAvailable
now check if either the client is an entitled web browser or the relying party
is in the client's associated domains and return false if neither are true.

Existing checks are performed after this new check.

To accomplish this, Document from BasicCredential is now plumbed through to
AuthenticatorCoordinator and SecurityOriginData is then passed to
WebAuthenticatorCoordinatorProxy.

* Source/WebCore/Modules/credentialmanagement/BasicCredential.cpp:
(WebCore::BasicCredential::isConditionalMediationAvailable):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::isConditionalMediationAvailable const):
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.h:
* Source/WebCore/Modules/webauthn/AuthenticatorCoordinatorClient.h:
* Source/WebKit/Platform/spi/Cocoa/AuthenticationServicesCoreSPI.h:
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/WebAuthenticatorCoordinatorProxy.mm:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.h:
* Source/WebKit/UIProcess/WebAuthentication/WebAuthenticatorCoordinatorProxy.messages.in:
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.cpp:
(WebKit::WebAuthenticatorCoordinator::isConditionalMediationAvailable):
* Source/WebKit/WebProcess/WebAuthentication/WebAuthenticatorCoordinator.h:

Canonical link: https://commits.webkit.org/259253@main
